plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/HudiSplitSource.java

```diff
@@ -76,14 +78,32 @@ public HudiSplitSource(
         List<HiveColumnHandle> partitionColumnHandles = table.getPartitionColumns().stream()
                 .map(column -> partitionColumnHandleMap.get(column.getName())).collect(toList());
 
-        HudiDirectoryLister hudiDirectoryLister = new HudiReadOptimizedDirectoryLister(
-                metadataConfig,
-                engineContext,
-                tableHandle,
-                metaClient,
-                metastore,
-                table,
-                partitionColumnHandles);
+        HudiTableType tableType = tableHandle.getTableType();
+        final HudiDirectoryLister hudiDirectoryLister;
+        switch (tableType) {
+            case COW:
+                hudiDirectoryLister = new HudiReadOptimizedDirectoryLister(
+                        metadataConfig,
+                        engineContext,
+                        tableHandle,
+                        metaClient,
+                        metastore,
+                        table,
+                        partitionColumnHandles);
+                break;
+            case MOR:
+                hudiDirectoryLister = new HudiRealTimeDirectoryLister(
+                        table,
+                        tableHandle,
+                        metastore,
+                        engineContext,
+                        metadataConfig,
+                        metaClient,
+                        partitionColumnHandles);
+                break;
```

Comment:
If you do intend to use `HudiTableType.UNKNOWN` then add that `case` as well. However, I would prefer if that element of enum is removed.